### PR TITLE
Add internal-metrics endpoint to gateway

### PR DESCRIPTION
**Pull request by mdubbyap**
Datapoints serialized to json available at configurable host:port/internal-metrics

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -417,6 +419,26 @@ func splitSinks(forwarders []protocol.Forwarder) ([]dpsink.DSink, []dpsink.ESink
 	return dsinks, esinks, tsinks
 }
 
+func (p *gateway) setupInternalMetricsServer(conf *config.GatewayConfig, logger log.Logger, scheduler *sfxclient.Scheduler) error {
+	if conf.InternalMetrics == nil {
+		return nil
+	}
+	listener, err := net.Listen("tcp", *conf.InternalMetrics)
+	if err != nil {
+		return errors.Annotate(err, "cannot setup prometheus metrics server")
```
Comment (mpetazzoni) [inline main.go]: This is no longer about Prometheus (same thing below)

File: config/config.go [hunk 2]
```diff
@@ -111,6 +111,7 @@ type GatewayConfig struct {
 	UnhealthyMemberTTL            *time.Duration    `json:"-"`
 	RemoveMemberTimeout           *time.Duration    `json:"-"`
 	AdditionalDimensions          map[string]string `json:",omitempty"`
+	InternalMetrics               *string           `json:",omitempty"`
```
Comment (mpetazzoni) [inline config/config.go]: Should this be `InternalMetricsListenAddr`? Current name sounds like it's a boolean config flag.

File: main_test.go [hunk 3]
```diff
@@ -196,6 +196,26 @@ const emptyConfig = `
...
+  }
+`
+
+const internalMetrics = `
+  {
+    "ListenFrom":[
+    ],
+    "ForwardTo":[
+    ],
+    "InternalMetrics": "0.0.0.0:0"
```
Comment (keitwb) [inline main_test.go]: Port 0 is valid I guess but it will be hard to know what port the kernel chooses since it never logs it out.

Comment (mdubbyap) [reply main_test.go]: That's that point. If i hard code it it's possible the test will fail.

File: main.go [hunk 4]
```diff
@@ -663,6 +692,11 @@ func (p *gateway) run(ctx context.Context) error {
 		return err
 	}
 
+	if err := p.setupInternalMetricsServer(loadedConfig, logger, scheduler); err != nil {
+		p.logger.Log(log.Err, "prometheus metrics server failed", err)
```
Comment (mpetazzoni) [inline main.go]: Missed that one

**Pull request merged**